src/Avalonia.Base/Utilities/DeferredSetter.cs

```diff
@@ -0,0 +1,148 @@
+        internal void AddPendingSet(TProperty property, TValue value)
+        {
+            Contract.Requires<InvalidOperationException>(IsNotifying(property));
+
+            setRecords.GetOrCreateValue(property).PendingValues.Enqueue(value);
+        }
+
+        /// <summary>
+        /// Checks if there are any pending assignments for the property.
+        /// </summary>
+        /// <param name="property">The property to check.</param>
+        /// <returns>If the property has any pending assignments.</returns>
+        internal bool HasPendingSet(TProperty property)
+        {
+            return setRecords.TryGetValue(property, out var status) && status.PendingValues.Count != 0;
+        }
+
+        /// <summary>
+        /// Gets the first pending assignment for the property.
+        /// </summary>
+        /// <param name="property">The property to check.</param>
+        /// <returns>The first pending assignment for the property.</returns>
+        internal TValue GetFirstPendingSet(TProperty property)
+        {
+            return setRecords.GetOrCreateValue(property).PendingValues.Dequeue();
+        }
+
+        /// <summary>
+        /// Set the property and notify listeners while ensuring we don't get into a stack overflow as happens with #855 and #824
+        /// </summary>
+        /// <param name="property">The property to set.</param>
+        /// <param name="setterCallback">
+        /// A callback that actually sets the property.
+        /// The first parameter is the value to set, and the second is a wrapper that takes a callback that sends the property-changed notification.
+        /// </param>
+        /// <param name="value">The value to try to set.</param>
+        /// <param name="pendingSetCondition">A predicate to filter what possible values should be added as pending sets (i.e. only values not equal to the current value).</param>
+        public void SetAndNotify(
+            TProperty property,
+            Action<TValue, Action<Action>> setterCallback,
+            TValue value,
+            Predicate<TValue> pendingSetCondition)
+        {
+            Contract.Requires<ArgumentNullException>(setterCallback != null);
+            if (!IsNotifying(property))
+            {
+                setterCallback(value, notification =>
+                {
+                    using (MarkNotifying(property))
+                    {
+                        notification();
+                    }
+                });
+                while (HasPendingSet(property))
+                {
+                    setterCallback(GetFirstPendingSet(property), notification =>
+                    {
+                        using (MarkNotifying(property))
+                        {
+                            notification();
+                        }
+                    });
+                }
+            }
+            else if(pendingSetCondition?.Invoke(value) ?? true)
```

Comment:
Why is `pendingSetNotification` passed in as `!object.Equals` by `PriorityValue` and `null` (i.e. always returns `true`) by `SetAndRaise`? Why the difference in these two cases?

Reply:
When I initially implementing this, for some reason using `!object.Equals` in all cases didn't work. I tried it again and it worked, so I'm making `pendingSetCondition` required and using `!object.Equals` in `SelectingItemsControl`.